libs/auth/src/services/pin-crypto/pin-crypto.service.spec.ts

```diff
@@ -0,0 +1,189 @@
+    );
+
+    function mockPinEncryptedKeyDataByPinLockType(
+      pinLockType: PinLockType,
+      migrationStatus: "PRE" | "POST" = "POST",
+    ) {
+      switch (pinLockType) {
+        case "PERSISTANT":
+          stateService.getPinKeyEncryptedUserKey.mockResolvedValue(
+            pinKeyEncryptedUserKeyPersistant,
+          );
+          if (migrationStatus === "PRE") {
+            stateService.getEncryptedPinProtected.mockResolvedValue(
+              oldPinKeyEncryptedMasterKeyPreMigrationPersistent,
+            );
+          } else {
+            stateService.getEncryptedPinProtected.mockResolvedValue(
+              oldPinKeyEncryptedMasterKeyPostMigration,
+            );
+          }
+          break;
+        case "TRANSIENT":
+          stateService.getPinKeyEncryptedUserKeyEphemeral.mockResolvedValue(
+            pinKeyEncryptedUserKeyEphemeral,
+          );
+
+          if (migrationStatus === "PRE") {
+            stateService.getDecryptedPinProtected.mockResolvedValue(
+              oldPinKeyEncryptedMasterKeyPreMigrationEphemeral,
+            );
+          } else {
+            stateService.getDecryptedPinProtected.mockResolvedValue(
+              oldPinKeyEncryptedMasterKeyPostMigration,
+            );
+          }
+          break;
+        case "DISABLED":
+          // no mocking required. Error should be thrown
+          break;
+      }
+    }
+
+    const testCases: { pinLockType: PinLockType; migrationStatus: "PRE" | "POST" }[] = [
+      { pinLockType: "PERSISTANT", migrationStatus: "PRE" },
+      { pinLockType: "PERSISTANT", migrationStatus: "POST" },
+      { pinLockType: "TRANSIENT", migrationStatus: "PRE" },
+      { pinLockType: "TRANSIENT", migrationStatus: "POST" },
+    ];
+
+    testCases.forEach(({ pinLockType, migrationStatus }) => {
+      it(`should successfully decrypt and return user key with valid, ${pinLockType} PIN (${migrationStatus} migration)`, async () => {
+        // Arrange
+        setupDecryptUserKeyWithPinMocks(pinLockType, migrationStatus);
+
+        // Act
+        const result = await pinCryptoService.decryptUserKeyWithPin(mockPin);
+
+        // Assert
+        expect(result).toEqual(mockUserKey);
+      });
+
+      it(`should return null when ${pinLockType} PIN (${migrationStatus} migration) is incorrect and user key cannot be decrypted`, async () => {
+        // Arrange
+        setupDecryptUserKeyWithPinMocks("PERSISTANT");
+
+        cryptoService.decryptUserKeyWithPin.mockResolvedValue(null);
+
+        // Act
+        const result = await pinCryptoService.decryptUserKeyWithPin(mockPin);
+
+        // Assert
+        expect(result).toBeNull();
+      });
+
+      // not sure if this is a realistic scenario but going to test it anyway
+      it(`should return null when ${pinLockType} PIN (${migrationStatus} migration) doesn't match after successful user key decryption`, async () => {
```

Comment:
nit: missing parenthesis again :D

suggestion: I think you are missing the word `pin` somewhere in here

```suggestion
      it(`should return null when the pin doesn't match after successful user key decryption`, async () => {
```

Reply:
There is an opening parenthesis. 

`should return null when ${pinLockType} PIN (${migrationStatus} migration) doesn't match after successful user key decryption` also includes the word PIN in the middle of the two interpolated values? 

Here's an example: "should return null when TRANSIENT **PIN** (PRE migration) doesn't match after successful user key decryption

Reply:
See https://github.com/bitwarden/clients/pull/7107#discussion_r1427456012 for full test run

Reply:
Yep, blind :D